### DISCUSSION TITLE
custom jupyter widget

### DISCUSSION THREAD

**Discussion by cornhundred** [jupyterlite/jupyterlite#222] [Q&A] [answered]
This is a great project! Is it possible to get a custom jupyter widget (e.g. our https://github.com/ismms-himc/clustergrammer2 project) working on Jupyterlite? I do not see how to install the notebook extension 

```
pip install clustergrammer2
jupyter nbextension install --py --sys-prefix clustergrammer2
jupyter nbextension enable --py --sys-prefix clustergrammer2
```

or Jupyter lab extension

```
jupyter labextension install @jupyter-widgets/jupyterlab-manager
```

**Comment by jtpio** [comment, marked as answer]
Thanks @cornhundred.

If it's a Jupyter Widget then it's very likely to work in JupyterLite as long as the Python dependencies can be installed in Pyodide.

You will first need to update to the new extension distribution system in JupyterLab 3.0: https://jupyterlab.readthedocs.io/en/stable/extension/extension_migration.html#jupyterlab-2-x-to-3-x

During a `jupyter lite build` the frontend extension should then be picked up: https://jupyterlite.readthedocs.io/en/latest/configuring.html#adding-extensions

Finally in the notebook users will have to install the Python package with:

```python
import micropip
await micropip.install('clustergrammer2')
```

**Reply by bollwyvl** [reply]
It is highly unlikely jupyterlite will _ever_ support "source" extensons, so the upstream play is the right path forward.

It is _possible_ to support both lab 2 and lab 3 distributions in one package, as [pythreejs](https://github.com/jupyter-widgets/pythreejs) does, but there isn't much carrot left for that, and the lab 2 mechanism was always kind of a farce, as it still download 100s of MBs of `npm` packages. I for one don't miss getting `webpack` OoM or `Path too long` or `too many files open` tracebacks posted as issues on "simple" widgets where folk are trying to install badly-behaved packages that inline huge dependencies.

**Reply by cornhundred** [reply]
Thanks @jtpio and @bollwyvl for the fast responses! I'll try that out and let you all know if it works.

**Reply by cornhundred** [reply]
@jtpio, this extension update applies to widgets too?

**Reply by jtpio** [reply]
@cornhundred yes, most widgets now distribute prebuilt extensions for JupyterLab 3.0.

Maybe the easiest could be to look at an existing widget, for example ipyleaflet: https://github.com/jupyter-widgets/ipyleaflet

**Reply by cornhundred** [reply]
Thanks, @jtpio. I made the changes described in the [extension migration guide](https://jupyterlab.readthedocs.io/en/stable/extension/extension_migration.html#upgrading-library-versions-manually) and the changes in the [updates for widget maintainers](https://ipywidgets.readthedocs.io/en/stable/changelog.html#updates-for-widget-maintainers) - see [package.json](https://github.com/ismms-himc/clustergrammer2/blob/master/package.json#L51). It seems to be working for Jupyter Lab 3 (version 3.0.3) on my machine and I'll update here if I can get it working on Jupyterlite

**Reply by cornhundred** [reply]
Although it required a rebuild and restart, so I probably did not set up the extension thing correctly.

**Reply by jtpio** [reply]
Looks like there might be a couple of things missing. For example the extension also needs to be placed under `share/jupyter/labextensions`.

**Reply by bollwyvl** [reply]
your going to need `jupyter labextension build` in that `package.json#/scripts` someplace... as well as `outputDir` in `#/jupyter/lab` to deploy it to... wherever you're going to put it

**Reply by jasongrout** [reply]
See also the typescript widget cookiecutter: https://github.com/jupyter-widgets/widget-ts-cookiecutter or the javascript widget cookiecutter: https://github.com/jupyter-widgets/widget-cookiecutter

**Reply by cornhundred** [reply]
Thanks @jasongrout, using the latest  https://github.com/jupyter-widgets/widget-ts-cookiecutter I was able to release a new version of Clustergrammer2@0.18.0 with the pre-built Jupyter lab extension. I also am able to get the widget working on a local build of Jupyter-Lite - thanks for your help @jtpio, @bollwyvl - it is very cool to see that working. I am working on deploying to GitHub pages using the demo repo - my branch is here https://github.com/cornhundred/clustergrammer2-lite. Is it possitlbe to just upload the results from a `jupyter lite build` to GitHub and enable github pages to deploy?

**Reply by cornhundred** [reply]
The demo appears to be working https://cornhundred.github.io/clustergrammer2-lite/lab/ - running the notebook `content/Clustergrammer2_Lite.ipynb` works thanks for the help!

**Reply by jtpio** [reply]
Congrats @cornhundred, this is neat!

**Reply by cornhundred** [reply]
FYI - it's also possible to run Pyodide on ObservableHQ. I was able to get this similar example workin https://observablehq.com/@cornhundred/clustergrammer2-pyodide
